### ISSUE TITLE
JSON Output - Enhancement

### ISSUE THREAD

**Issue by Reloc** [szpajder/dumpvdl2#4]
Hi Tomasz,

Would it be possible to add an additional output to dumpvdl2 to send JSON via UDP to an IP address ?
vdlm2dec already has this function & its very useful as the output to Planeplotter is limited in what it supplies.

JSON output would include Date, Time, Channel, Frequency, Aircraft ICAO, Ground Station Id, Direction/Uplink/Downlink indicator, Latitude, Longitude, Altitude and Destination.

Regards
Nigel

**Comment by szpajder** [comment]
This is on TODO, but I don't want this to be limited to just a single type
of frame and a handful of params. I want to JSONize all parameters decoded
and displayed by the program from all supported packet types. But it's
quite a huge task which needs the output layer code to be restructured
first, so I can't give you an ETA for this feature yet.

**Comment by kevinelliott** [comment]
This would be great!

**Comment by TanerH** [comment]
Ooh, JSON output would be awesome -- even if it's just to a(nother) log file

**Comment by kevinelliott** [comment]
So... something like this? 

```
pi@hambox1:~/dumpvdl2-custom/build $ sudo dumpvdl2 --rtlsdr 0 --output-format-json --centerfreq 136955000 136975000 136725000
dumpvdl2 v1.6.0-20-g99c332d (libacars 1.1.0)
Sampling rate set to 1050000 sps
Found 1 device(s):
  0:  Realtek, RTL2832U, SN: fa-amp

Using device 0: Generic RTL2832U
Found Rafael Micro R820T tuner
Exact sample rate is: 1050000.026077 Hz
[R82XX] PLL not locked!
Center frequency set to 136955000 Hz
Device #0: gain set to automatic
Device 0 started


{ "time": "2019-02-13 11:15:50 GMT", "freq": 136.975, "signal_power_dbfs": -20.0, "noise_floor_power_dbfs": -37.8, "net_dbfs": 17.8, "ppm": 0.7, "source_address": "AD67CC", "source_type": "Aircraft", "source_status": "Airborne", "destination_address": "278578", "destination_type": "Ground station", "destination_status": "Airborne", "avlc": { "type": "I", "sseq": 1, "rseq": 2, "poll": 0, "proto": "acars", "acars": { "mode": "2", "reg": ".N963FD", "ack": "G", "label": "_d", "block_id": "2", "no": "S42A", "flight_id": "FX1859", "txt": "" } } }
{ "time": "2019-02-13 11:15:51 GMT", "freq": 136.975, "signal_power_dbfs": -19.9, "noise_floor_power_dbfs": -38.1, "net_dbfs": 18.2, "ppm": 0.8, "source_address": "AD67CC", "source_type": "Aircraft", "source_status": "Airborne", "destination_address": "278578", "destination_type": "Ground station", "destination_status": "Airborne", "avlc": { "type": "I", "sseq": 2, "rseq": 2, "poll": 0, "proto": "acars", "acars": { "mode": "2", "reg": ".N963FD", "ack": "!", "label": "H1", "block_id": "3", "no": "F30A", "flight_id": "FX1859", "txt": "#M2B/BA USADCXA.DR1.N963FD2056" } } }
{ "time": "2019-02-13 11:15:53 GMT", "freq": 136.975, "signal_power_dbfs": -20.4, "noise_floor_power_dbfs": -37.8, "net_dbfs": 17.4, "ppm": 0.8, "source_address": "AD67CC", "source_type": "Aircraft", "source_status": "Airborne", "destination_address": "278578", "destination_type": "Ground station", "destination_status": "On ground", "avlc": { "type": "S", "sfunc": "Receive Ready", "p/f": 0, "rseq": 3 } }
{ "time": "2019-02-13 11:16:13 GMT", "freq": 136.975, "signal_power_dbfs": -18.8, "noise_floor_power_dbfs": -38.4, "net_dbfs": 19.6, "ppm": 0.9, "source_address": "AD67CC", "source_type": "Aircraft", "source_status": "Airborne", "destination_address": "278578", "destination_type": "Ground station", "destination_status": "Airborne", "avlc": { "type": "I", "sseq": 3, "rseq": 3, "poll": 0, "proto": "acars", "acars": { "mode": "2", "reg": ".N963FD", "ack": "!", "label": "H1", "block_id": "4", "no": "D63A", "flight_id": "FX1859", "txt": "#DFB5WT4021311160096-128174806551771200414  1114290379827-1218084017480-1282330650  1114540380253-1217643018633-1532340660  1115290380857-1217011020353-1802410690  1116040381473-1216366021770-2152400740  " } } }
{ "time": "2019-02-13 11:16:15 GMT", "freq": 136.975, "signal_power_dbfs": -18.2, "noise_floor_power_dbfs": -38.1, "net_dbfs": 19.8, "ppm": 0.9, "source_address": "AD67CC", "source_type": "Aircraft", "source_status": "Airborne", "destination_address": "278578", "destination_type": "Ground station", "destination_status": "On ground", "avlc": { "type": "S", "sfunc": "Receive Ready", "p/f": 0, "rseq": 4 } }
{ "time": "2019-02-13 11:16:47 GMT", "freq": 136.975, "signal_power_dbfs": -18.0, "noise_floor_power_dbfs": -38.3, "net_dbfs": 20.3, "ppm": 0.7, "source_address": "AD67CC", "source_type": "Aircraft", "source_status": "Airborne", "destination_address": "278578", "destination_type": "Ground station", "destination_status": "Airborne", "avlc": { "type": "I", "sseq": 4, "rseq": 4, "poll": 0, "proto": "acars", "acars": { "mode": "2", "reg": ".N963FD", "ack": "!", "label": "H1", "block_id": "5", "no": "D64A", "flight_id": "FX1859", "txt": "#DFB5TKM021311160005-255230377241764800414  315NOT_TO????????  091108740900  091108830917  095908980910  095409100928  " } } }
{ "time": "2019-02-13 11:16:48 GMT", "freq": 136.975, "signal_power_dbfs": -17.6, "noise_floor_power_dbfs": -37.7, "net_dbfs": 20.1, "ppm": 0.8, "source_address": "AD67CC", "source_type": "Aircraft", "source_status": "Airborne", "destination_address": "278578", "destination_type": "Ground station", "destination_status": "On ground", "avlc": { "type": "S", "sfunc": "Receive Ready", "p/f": 0, "rseq": 5 } }
{ "time": "2019-02-13 11:17:30 GMT", "freq": 136.975, "signal_power_dbfs": -16.8, "noise_floor_power_dbfs": -38.5, "net_dbfs": 21.7, "ppm": 0.7, "source_address": "AD67CC", "source_type": "Aircraft", "source_status": "Airborne", "destination_address": "2939F7", "destination_type": "Ground station", "destination_status": "Airborne", "avlc": { "type": "U", "mfunc": "XID", "p/f": 1 } }
{ "time": "2019-02-13 11:17:38 GMT", "freq": 136.975, "signal_power_dbfs": -15.1, "noise_floor_power_dbfs": -37.7, "net_dbfs": 22.5, "ppm": 0.7, "source_address": "AD67CC", "source_type": "Aircraft", "source_status": "Airborne", "destination_address": "2939F7", "destination_type": "Ground station", "destination_status": "Airborne", "avlc": { "type": "U", "mfunc": "XID", "p/f": 1 } }
{ "time": "2019-02-13 11:17:41 GMT", "freq": 136.975, "signal_power_dbfs": -15.9, "noise_floor_power_dbfs": -37.8, "net_dbfs": 22.0, "ppm": 0.9, "source_address": "AD67CC", "source_type": "Aircraft", "source_status": "Airborne", "destination_address": "278578", "destination_type": "Ground station", "destination_status": "Airborne", "avlc": { "type": "I", "sseq": 5, "rseq": 5, "poll": 0, "proto": "acars", "acars": { "mode": "2", "reg": ".N963FD", "ack": "!", "label": "H1", "block_id": "6", "no": "D65A", "flight_id": "FX1859", "txt": "#DFB5WT4021311170004-256231397271764800414  1116390382105-1215705023139-2562390710  1117140382742-1215037024669-2842380640  1117290383011-1214754024968-2892380610  1117320383064-1214697025002-2902390610  " } } }
{ "time": "2019-02-13 11:17:42 GMT", "freq": 136.975, "signal_power_dbfs": -15.3, "noise_floor_power_dbfs": -38.2, "net_dbfs": 22.9, "ppm": 0.7, "source_address": "AD67CC", "source_type": "Aircraft", "source_status": "Airborne", "destination_address": "278578", "destination_type": "Ground station", "destination_status": "On ground", "avlc": { "type": "S", "sfunc": "Receive Ready", "p/f": 0, "rseq": 6 } }
{ "time": "2019-02-13 11:17:47 GMT", "freq": 136.975, "signal_power_dbfs": -15.0, "noise_floor_power_dbfs": -37.6, "net_dbfs": 22.6, "ppm": 0.9, "source_address": "AD67CC", "source_type": "Aircraft", "source_status": "Airborne", "destination_address": "2939F7", "destination_type": "Ground station", "destination_status": "Airborne", "avlc": { "type": "U", "mfunc": "XID", "p/f": 1 } }
{ "time": "2019-02-13 11:17:56 GMT", "freq": 136.975, "signal_power_dbfs": -14.6, "noise_floor_power_dbfs": -37.7, "net_dbfs": 23.1, "ppm": 0.8, "source_address": "AD67CC", "source_type": "Aircraft", "source_status": "Airborne", "destination_address": "2939F7", "destination_type": "Ground station", "destination_status": "Airborne", "avlc": { "type": "U", "mfunc": "XID", "p/f": 1 } }
{ "time": "2019-02-13 11:17:57 GMT", "freq": 136.975, "signal_power_dbfs": -15.0, "noise_floor_power_dbfs": -37.9, "net_dbfs": 22.9, "ppm": 0.9, "source_address": "AD67CC", "source_type": "Aircraft", "source_status": "Airborne", "destination_address": "2939F7", "destination_type": "Ground station", "destination_status": "Airborne", "avlc": { "type": "I", "sseq": 0, "rseq": 0, "poll": 0, "proto": "acars", "acars": { "mode": "2", "reg": ".N963FD", "ack": "!", "label": "Q0", "block_id": "7", "no": "S43A", "flight_id": "FX1859", "txt": "" } } }
{ "time": "2019-02-13 11:17:58 GMT", "freq": 136.975, "signal_power_dbfs": -14.4, "noise_floor_power_dbfs": -37.4, "net_dbfs": 23.0, "ppm": 1.0, "source_address": "AD67CC", "source_type": "Aircraft", "source_status": "Airborne", "destination_address": "2939F7", "destination_type": "Ground station", "destination_status": "On ground", "avlc": { "type": "S", "sfunc": "Receive Ready", "p/f": 0, "rseq": 1 } }
```

😄

**Comment by Reloc** [comment]
Nice work Kevin thanks :) Now if you can just get the output going to UDP then it would be even better ;)

Regards
Nigel

**Comment by kevinelliott** [comment]
Output over UDP in which format? I could add another format and then you could pipe the output using netcat/socat until some networking was added to the lib.

**Comment by Reloc** [comment]
It would ideally be JSON format , I have it kind of working with

 dumpvdl2 --rtlsdr 0 --output-format-json --centerfreq 136955000 136975000 136725000 | nc -u 192.168.xxx.xxx 12345

Maybe I just can't see it but would it be possible to open "issues" on your Fork to save spamming this builds thread ;)

Regards
Nigel

**Comment by kevinelliott** [comment]
@Reloc Enabled `Issues` on my fork! Eventually I will submit a PR to this repo, but I think mine will make a good working area as we sort through the issues on my work.

BTW, good thing you hid your internal IP address in that command you pasted 😝

**Comment by szpajder** [comment]
@kevinelliott thanks for your work, but please be aware that the output generation code in dumpvdl2 will be completely replaced with a new implementation in the next release (or two), so whatever you do with the current code will not be suitable for merging back.

**Comment by kevinelliott** [comment]
@szpajder Alright, hate to call this out, but over 1 year later and still no JSON output. Not trying to be a nag, but figured 1 year is probably a good place to start the dialog again. Is your new implementation on the forefront or should I maintain a separate branch for others to get JSON output?

**Comment by szpajder** [comment]
JSON is on TODO, however I wanted to complete the work on all decoders first, so that I could then add JSON support to all of them in one go instead of going back to this topic multiple times and possibly changing the JSON structure several times in a backward-incompatible way, whenever a major change is done to any of the decoders.

When I completed this step, then ADS-C v2 went on air, so I put everything else aside to implement a decoder for it (and I was quite in a rush, because it is said to be a trial run, so not sure how long it's going to last). And when it was ready, it appeared that it doesn't make sense to have ADS-C v2 decoding without message reassembly support, because ADS-C v2 messages are so large that most of them are split into multiple fragments and the decoder was unable to deal with fragmented messages. And again, I might have done it in a quick'n'messy way and only make it able to deal with X.25 traffic, but given the fact that ACARS messages are also fragmented, I would implement ACARS reassembly eventually anyway, so to avoid code duplication, I decided to make this in the right and generic way, and this is how it's been implemented. But it took some more time.

So it's not that the features drop off the TODO list. Things just get prioritized or deprioritized, sometimes due to my liking and other times due to external circumstances (like ADS-C v2).

And speaking about last year - there were 7 dumpvdl2 releases and 5 libacars releases during that time. No benefit from the new features for you?

**Comment by kevinelliott** [comment]
@szpajder Thanks for the update, and yes, that tracks well with what I've been seeing.

As for benefit for me -- no, unfortunately the only thing that benefits me is JSON output (see: https://app.airframes.io) -- since I aggregate data and spend my time making meaningful sense of the data, rather than watch it scroll by in a format that is not easily parsed. So, that said, I have been using vdlm2dec instead of dumpvdl2 since it does have JSON output.

But the many improvements to dumpvdl2 look incredible. There are several of us excited to jump on and use it once it has the output.

Cheers!

**Comment by szpajder** [comment]
JSON formatting is now supported in unstable branch of dumpvdl2.

**Comment by szpajder** [comment]
Released in v2.1.0.

**Comment by kevinelliott** [comment]
Just a heads up, the dumpvdl2 feeder ingest is active at https://app.airframes.io/about ! 

Hostname: `feed.airframes.io`
Protocol: `TCP`
Port: `5553`

If for some reason someone wants to use UDP:

Hostname: `feed.airframes.io`
Protocol: `UDP`
Port: `5552`

Though I would encourage TCP since it's more reliable and there's less change that a larger message will get clipped.

Be sure to specify a station id with `--station-id` such as `KE-KMHR-VDL1` -- I put some general suggestions on the page linked above.
